### PR TITLE
Setup wizard: add a browser step

### PR DESCRIPTION
**Pull request by qparis**

### CODE REVIEW

**Comment by qparis** [comment]
It behaves like this: 

```
sw = SetupWizard(InstallationType.APPS, "Test", java.util.Optional.empty());
browserWindow =  sw.createBrowser("Hello");
browserWindow.goToUrl("https://auth.gog.com/auth?client_id=46899977096215655&redirect_uri=https%3A%2F%2Fembed.gog.com%2Fon_login_success%3Forigin%3Dclient&response_type=code&layout=client2");
browserWindow.waitForUrl("https://embed.gog.com/*");
browserWindow.getCurrentUrl();
```

Will show: 

capture d ecran 2018-12-31 a 11 30 52

And will also wait the user to be logged in to continue. We can then retrieve the tokens from GoG.com (for example)

**Comment by madoar** [comment]
@qparis can you give me a more detailed example where this is needed? (I can't think of one, but I think this feature is cool)

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.phoenicis.scripts.ui.Message;
+
+import java.util.concurrent.Semaphore;
+
+public class StepRepresentationBrowser extends StepRepresentationMessage implements BrowserControl {
+    private final WebView webView = new WebView();
+    private final Message<?> messageWaitingForResponse;
+
+    public StepRepresentationBrowser(SetupUiJavaFXImplementation parent, Message<?> messageWaitingForResponse,
+            String textToShow) {
```
[the remark this answers is not in the document]
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java]: I also think the class name is strange.
How about using `BrowserStep` or `BrowserRepresentationStep`?

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        try {
+            lock.acquire();
+        } catch (InterruptedException e) {
+            this.messageWaitingForResponse.sendCancelSignal();
+        }
+
+    }
+
+    private boolean urlMatches(String currentUrl, String urlMatch) {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java]: JavaDoc

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/BrowserControl.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package org.phoenicis.scripts.ui;
+
+public interface BrowserControl {
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/BrowserControl.java]: JavaDoc

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/SetupUi.java [hunk 4]
```diff
@@ -83,6 +83,12 @@
      */
     void showProgressBar(Message<ProgressControl> message, String textToShow);
 
+    /**
+     * Shows a browser inside the window and returns immediately
+     * @param textToShow The text to show to the user
+     */
+    void showBrowser(Message<BrowserControl> doneCallback, String textToShow);
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/SetupUi.java]: JavaDoc for `doneCallback` is missing

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/BrowserWizard.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.phoenicis.scripts.wizard;
+
+import org.phoenicis.scripts.ui.BrowserControl;
+
+public interface BrowserWizard {
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/BrowserWizard.java]: JavaDoc

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/SetupWizardBrowser.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package org.phoenicis.scripts.wizard;
+
+interface SetupWizardBrowser extends Wizard {
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/SetupWizardBrowser.java]: JavaDoc + the interface name sounds weird, maybe better `BrowserSetupWizard`?

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/WineWizard.java [hunk 7]
```diff
@@ -18,7 +18,7 @@
 
 package org.phoenicis.scripts.wizard;
 
-public interface WineWizard extends ChoiceWizard, ProgressWizard, LogWizard {
+public interface WineWizard extends ChoiceWizard, ProgressWizard, LogWizard, BrowserWizard {
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/WineWizard.java]: JavaDoc

**Comment by qparis** [comment]
> @qparis can you give me a more detailed example where this is needed? (I can't think of one, but I think this feature is cool)

This should allow scripts to download GOG games. This would lead to potentially tons of scripts. I’d like to cherry pick this feature for alpha 2 if possible because it was a really appreciated feature in POLv4, it is now broken.

 I think that the community will appreciate it

**Comment by qparis** [comment]
(I promess, it is the last one)

**Comment by madoar** [comment]
Don't make promises you can't keep :D

**Comment by madoar** [comment]
To use this feature with GOG we should also provide a `GOGScript`, similar to the `SteamScript` and the `OriginScript`, right?

**Comment by qparis** [comment]
> To use this feature with GOG we should also provide a `GOGScript`, similar to the `SteamScript` and the `OriginScript`, right?

Ideally yes. I just need to check that it is possible to download the setup file with the game but it should be possible by some way

**Comment by plata** [comment]
I don't think we should take yet another feature for alpha-2. We're in a continuous improvement and features are added more or less regularly so I don't see the point to add this (especially if we have now real scripts for it like @madoar said). We have to stop at some point otherwise we will never get alpha-2 released. Everything that's not in now will be in for the next release (we don't have to wait 1/2 year do do that).

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java]: The class name follows the pattern of the other classe namess. By the way, it is easier to put the interface before the name of the class, it makes implementations quicker to find

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java (hunk 2 above)
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java]: Normally, it is not needed for private methods but I'll add it

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/scriptui/StepRepresentationBrowser.java]: I'm accustomed to the representation form, where the interface name is used at the end of class name.
This also seems to be the way Java handles it with its own classes/interfaces

**Comment by madoar** [comment]
I'm fine either releasing this alpha 2 or 3.

**Comment by plata** [comment]
@qparis if you say so... However, I insist that we we think about stuff like this next time **before** we decide to create the release branch. Also, please ensure that everything is still working after merging this (by following the test plan in the documentation).

**Comment by plata** [comment]
@madoar are all your remarks fixed so that this can be merged?

**Comment by qparis** [comment]
I think they are

**Comment by plata** [comment]
@qparis please check again. It seems indeed that some of them are still not fixed.

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/SetupUi.java (hunk 4 above)
Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/ui/SetupUi.java]: It is missing for all the methods. Maybe we should handle this in a separate ticket

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/SetupWizardBrowser.java (hunk 6 above)
Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/SetupWizardBrowser.java]: It was not used

**Comment by qparis** [comment]
I think we are good now.

**Comment by plata** [comment]
Can be merged.

**Pull request merged**